File: Main/conversion.js
const fs = require('fs');
const path = require('path');

/**
 * Parses a Newick string into a nested JavaScript object.
 */
function parseNewick(newick) {
    let ancestors = [];
    let tree = {};
    let tokens = newick.split(/\s*(;|\(|\)|,|:)\s*/);
    let subtree;

    for (let i = 0; i < tokens.length; i++) {
        let token = tokens[i];
        switch (token) {
            case '(':
                subtree = {};
                if (!tree.children) tree.children = [];
                tree.children.push(subtree);
                ancestors.push(tree);
                tree = subtree;
                break;
            case ',':
                subtree = {};
                ancestors[ancestors.length - 1].children.push(subtree);
                tree = subtree;
                break;
            case ')':
                tree = ancestors.pop();
                break;
            case ':':
                break;
            default:
                if (tokens[i - 1] === ':') {
                    tree.dist = parseFloat(token);
                } else if (token !== ';') {
                    tree.name = token;
                }
        }
    }
    return tree;
}

/**
 * Recursively finds the maximum depth of the tree.
 */
function getMaxDepth(node, depth = 1) {
    if (!node.children || node.children.length === 0) return depth;
    return Math.max(...node.children.map(child => getMaxDepth(child, depth + 1)));
}

/**
 * Formats distance values by trimming trailing zeros and decimal points.
 */
function formatDistance(dist) {
    return dist ? parseFloat(dist).toFixed(9).replace(/0+$/, '').replace(/\.$/, '') : '0';
}

/**
 * Converts a parsed Newick tree to a custom flat JSON format with node info.
 */
function newickToCustomFormat(tree) {
    const jsonNodes = [];

    // Add dummy root node for visualization anchoring
    jsonNodes.push({ id: 'root', name: 'Root', level: 0 });

    let nodeCounter = 0;

    function createNodeId() {
        return `Internal_node${nodeCounter++}`;
    }

    const maxDepth = getMaxDepth(tree, 2);

    function formatLabel(label) {
        if (!label) return '1';
        label = label.trim();
        if (!isNaN(label)) {
            const num = parseFloat(label);
            return Number.isInteger(num) ? num.toString() : num.toFixed(9).replace(/0+$/, '').replace(/\.$/, '');
        }
        return label;
    }

    function processNode(node, parentId, currentLevel) {
        if (!node.children || node.children.length === 0) {
            jsonNodes.push({
                id: node.name,
                parent: parentId,
                name: formatLabel(node.name),
                customLabel: formatDistance(node.dist),
                level: maxDepth,
            });
        } else {
            const currentId = createNodeId();
            const label = formatLabel(node.name);

            jsonNodes.push({
                id: currentId,
                parent: parentId,
                name: label,
                customLabel: formatDistance(node.dist),
                level: currentLevel,
            });

            node.children.forEach(child => processNode(child, currentId, currentLevel + 1));
        }
    }

    const rootId = createNodeId();
    const rootLabel = formatLabel(tree.name);

    jsonNodes.push({
        id: rootId,
        parent: 'root', // Link real tree to dummy root
        name: rootLabel,
        customLabel: formatDistance(tree.dist),
        level: 1,
    });

    tree.children?.forEach(child => processNode(child, rootId, 2));

    return jsonNodes;
}


// === Main Execution ===

// 👇 Replace with your input file path
const inputPath = "C:/Users/amgot/OneDrive/Desktop/Edgar Internship/EDGAR_Test/Main/newicks/EDGAR_Acidovorax_fasttree.newick";
const outputPath = inputPath.replace(".newick", ".json");

// Check if input file exists
if (!fs.existsSync(inputPath)) {
    console.error("❌ File not found. Please check the input path.");
    process.exit(1);
}

// Read and convert the Newick file
const newickData = fs.readFileSync(inputPath, 'utf8');
const parsedTree = parseNewick(newickData);
const jsonResult = newickToCustomFormat(parsedTree);

// Save as .json file
fs.writeFileSync(outputPath, JSON.stringify(jsonResult, null, 2), 'utf8');

console.log("✅ Tree converted to custom JSON format:");
console.log("📁 Saved to:", outputPath);

File: Main/conversion.py
import json
import re
from pathlib import Path

def parse_newick(newick_str):
    tokens = re.split(r'\s*(;|\(|\)|,|:)\s*', newick_str)
    tokens = [t for t in tokens if t not in ['', ';']]
    tree = {}
    ancestors = []
    subtree = None

    i = 0
    while i < len(tokens):
        token = tokens[i]
        if token == '(':
            subtree = {}
            tree.setdefault('children', []).append(subtree)
            ancestors.append(tree)
            tree = subtree
        elif token == ',':
            subtree = {}
            ancestors[-1].setdefault('children', []).append(subtree)
            tree = subtree
        elif token == ')':
            tree = ancestors.pop()
        elif token == ':':
            pass
        else:
            prev_token = tokens[i - 1] if i > 0 else None
            if prev_token == ':':
                tree['dist'] = float(token)
            else:
                tree['name'] = token
        i += 1
    return tree

def get_max_depth(node, depth=1):
    if 'children' not in node or not node['children']:
        return depth
    return max(get_max_depth(child, depth + 1) for child in node['children'])

def format_distance(dist):
    if dist is None:
        return '0'
    s = f"{float(dist):.9f}".rstrip('0').rstrip('.')
    return s or '0'

def format_label(label):
    if not label:
        return '1'
    label = str(label).strip()
    try:
        num = float(label)
        return str(int(num)) if num.is_integer() else format_distance(num)
    except ValueError:
        return label

def newick_to_custom_format(tree):
    json_nodes = [{'id': 'root', 'name': 'Root', 'level': 0}]
    node_counter = [0]

    def create_node_id():
        node_counter[0] += 1
        return f"Internal_node{node_counter[0] - 1}"

    max_depth = get_max_depth(tree, 2)

    def process_node(node, parent_id, current_level):
        if 'children' not in node or not node['children']:
            json_nodes.append({
                'id': node.get('name'),
                'parent': parent_id,
                'name': format_label(node.get('name')),
                'customLabel': format_distance(node.get('dist')),
                'level': max_depth
            })
        else:
            current_id = create_node_id()
            json_nodes.append({
                'id': current_id,
                'parent': parent_id,
                'name': format_label(node.get('name')),
                'customLabel': format_distance(node.get('dist')),
                'level': current_level
            })
            for child in node['children']:
                process_node(child, current_id, current_level + 1)

    root_id = create_node_id()
    json_nodes.append({
        'id': root_id,
        'parent': 'root',
        'name': format_label(tree.get('name')),
        'customLabel': format_distance(tree.get('dist')),
        'level': 1
    })

    for child in tree.get('children', []):
        process_node(child, root_id, 2)

    return json_nodes

# === MAIN EXECUTION ===

input_file = "newicks/EDGAR_Acidovorax_fasttree.newick"  # Replace with your file
output_file = input_file.replace(".newick", ".json")

if not Path(input_file).exists():
    print(f"❌ File not found: {input_file}")
else:
    with open(input_file, 'r') as f:
        newick_string = f.read()

    parsed = parse_newick(newick_string)
    result = newick_to_custom_format(parsed)

    with open(output_file, 'w') as f:
        json.dump(result, f, indent=2)

    print(f"✅ Conversion complete! JSON saved to: {output_file}")
